### ISSUE TITLE
ICE: MIR const-checker found novel structural match violation

### ISSUE THREAD

**Issue by jamesbornholt** [rust-lang/rust#73431]
This code causes an ICE on both beta and nightly, but not stable. bisect-rustc pointed me to #67343.

It seems to be an issue with the generic `Wrapper<T>` -- manually implementing `Zero` for `Wrapper<usize>` doesn't trigger the issue.

### Code

```Rust
pub trait Zero {
    const ZERO: Self;
}

impl Zero for usize {
    const ZERO: Self = 0;
}

impl<T: Zero> Zero for Wrapper<T> {
    const ZERO: Self = Wrapper(T::ZERO);
}

#[derive(Debug, PartialEq, Eq)]
pub struct Wrapper<T>(T);

fn is_zero(x: Wrapper<usize>) -> bool {
    match x {
        Zero::ZERO => true,
        _ => false,
    }
}
```

### Meta
`rustc --version --verbose`:
```
rustc 1.45.0-beta.2 (1dc0f6d8e 2020-06-15)
binary: rustc
commit-hash: 1dc0f6d8ef4ff19cfbe468ede40a1a6596f48957
commit-date: 2020-06-15
host: x86_64-apple-darwin
release: 1.45.0-beta.2
LLVM version: 10.0
```

### Error output

```
error: internal compiler error: src/librustc_mir_build/hair/pattern/const_to_pat.rs:111: MIR const-checker found novel structural match violation

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:907:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.45.0-beta.2 (1dc0f6d8e 2020-06-15) running on x86_64-apple-darwin

note: compiler flags: -C embed-bitcode=no -C debuginfo=2 -C incremental --crate-type lib
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
stack backtrace:
   0: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
   1: core::fmt::write
   2: std::io::Write::write_fmt
   3: std::panicking::default_hook::{{closure}}
   4: std::panicking::default_hook
   5: rustc_driver::report_ice
   6: std::panicking::rust_panic_with_hook
   7: std::panicking::begin_panic
   8: rustc_errors::HandlerInner::bug
   9: rustc_errors::Handler::bug
  10: rustc_middle::util::bug::opt_span_bug_fmt::{{closure}}
  11: rustc_middle::ty::context::tls::with_opt::{{closure}}
  12: rustc_middle::ty::context::tls::with_opt
  13: rustc_middle::util::bug::opt_span_bug_fmt
  14: rustc_middle::util::bug::bug_fmt
  15: rustc_mir_build::hair::pattern::const_to_pat::ConstToPat::to_pat
  16: rustc_infer::infer::InferCtxtBuilder::enter
  17: rustc_mir_build::hair::pattern::PatCtxt::lower_path
  18: rustc_mir_build::hair::pattern::PatCtxt::lower_pattern
  19: rustc_mir_build::hair::pattern::check_match::MatchVisitor::lower_pattern
  20: <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
  21: <rustc_mir_build::hair::pattern::check_match::MatchVisitor as rustc_hir::intravisit::Visitor>::visit_expr
  22: <rustc_mir_build::hair::pattern::check_match::MatchVisitor as rustc_hir::intravisit::Visitor>::visit_expr
  23: rustc_mir_build::hair::pattern::check_match::check_match
  24: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::check_match>::compute
  25: rustc_middle::dep_graph::<impl rustc_query_system::dep_graph::DepKind for rustc_middle::dep_graph::dep_node::DepKind>::with_deps
  26: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  27: rustc_data_structures::stack::ensure_sufficient_stack
  28: rustc_query_system::query::plumbing::get_query_impl
  29: rustc_query_system::query::plumbing::ensure_query_impl
  30: rustc_session::utils::<impl rustc_session::session::Session>::time
  31: rustc_session::utils::<impl rustc_session::session::Session>::time
  32: rustc_interface::passes::analysis
  33: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::analysis>::compute
  34: rustc_middle::dep_graph::<impl rustc_query_system::dep_graph::DepKind for rustc_middle::dep_graph::dep_node::DepKind>::with_deps
  35: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  36: rustc_query_system::query::plumbing::get_query_impl
  37: rustc_middle::ty::context::tls::enter_global
  38: rustc_interface::interface::run_compiler_in_existing_thread_pool
  39: rustc_ast::attr::with_globals

```

</p>
</details>

**Comment by lcnr** [comment]
cc @ecstatic-morse

**Comment by lcnr** [comment]
so afaict, the `mir_const_qualif` checks the polymorphic anonymous const `for<T: Zero>: Wrapper(T::ZERO)`, which is correctly identified as not being structurally match while `search_for_structural_match_violation` is run on the fully monomorphized `Wrapper<usize>`, which is correctly identified as structural match.

So in a sense `custum_eq` is too conservative here, don't really know how to fix this without just disabling this check

**Comment by Dylan-DPC** [comment]
Marking it as p-critical as per our [discussion](https://rust-lang.zulipchat.com/#narrow/stream/227806-t-compiler.2Fwg-prioritization/topic/I-prioritize.20.2373431.20ICE.3A.20MIR.20const-checker.20found.20novel.20struc)

**Comment by RalfJung** [comment]
So speaking of structure match/equality, was there any progress in documenting what the exact guarantees are here? Withe each bug like this, I feel more unease around the lack of documentation here.^^ And that is even more so when two checks for that property disagree...

The [previous Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/Terminology.20.22structural.20match.22) did not really reach any conclusion.

**Comment by LeSeulArtichaut** [comment]
Since #73446 was beta-backported, I will remove the critical priority from this issue.

**Comment by LeSeulArtichaut** [comment]
Assigning `P-medium` [as discussed here](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/I-prioritize.20.2373431.20ICE.3A.20MIR.20const-checker.20found.20novel.20struc).

**Comment by lcnr** [comment]
Tracking the remaining work here in #73448, closing this issue
